cli/src/commands/diff.rs

```diff
@@ -0,0 +1,437 @@
+    let entry_2_metadata: Metadata =
+        RuntimeMetadataPrefixed::decode(&mut &bytes[..])?.try_into()?;
+
+    Ok((entry_1_metadata, entry_2_metadata))
+}
+
+fn storage_differences<'a>(
+    pallet_metadata_1: &'a PalletMetadata<'a>,
+    pallet_metadata_2: &'a PalletMetadata<'a>,
+) -> Vec<Diff<&'a StorageEntryMetadata>> {
+    diff(
+        pallet_metadata_1
+            .storage()
+            .map(|s| s.entries())
+            .unwrap_or_default(),
+        pallet_metadata_2
+            .storage()
+            .map(|s| s.entries())
+            .unwrap_or_default(),
+        |e| {
+            pallet_metadata_1
+                .storage_hash(e.name())
+                .expect("storage entry should be present")
+        },
+        |e| {
+            pallet_metadata_2
+                .storage_hash(e.name())
+                .expect("storage entry should be present")
+        },
+        |e| e.name(),
+    )
+}
+
+fn calls_differences<'a>(
+    pallet_metadata_1: &'a PalletMetadata<'a>,
+    pallet_metadata_2: &'a PalletMetadata<'a>,
+) -> Vec<Diff<&'a Variant<PortableForm>>> {
+    return diff(
+        pallet_metadata_1.call_variants().unwrap_or_default(),
+        pallet_metadata_2.call_variants().unwrap_or_default(),
+        |e| {
+            pallet_metadata_1
+                .call_hash(&e.name)
+                .expect("call should be present")
+        },
+        |e| {
+            pallet_metadata_2
+                .call_hash(&e.name)
+                .expect("call should be present")
+        },
+        |e| &e.name,
+    );
+}
+
+fn constants_differences<'a>(
+    pallet_metadata_1: &'a PalletMetadata<'a>,
+    pallet_metadata_2: &'a PalletMetadata<'a>,
+) -> Vec<Diff<&'a ConstantMetadata>> {
+    diff(
+        pallet_metadata_1.constants(),
+        pallet_metadata_2.constants(),
+        |e| {
+            pallet_metadata_1
+                .constant_hash(e.name())
+                .expect("constant should be present")
+        },
+        |e| {
+            pallet_metadata_2
+                .constant_hash(e.name())
+                .expect("constant should be present")
+        },
+        |e| e.name(),
+    )
+}
+
+fn runtime_api_differences<'a>(
+    metadata_1: &'a Metadata,
+    metadata_2: &'a Metadata,
+) -> Vec<Diff<RuntimeApiMetadata<'a>>> {
+    diff(
+        metadata_1.runtime_api_traits(),
+        metadata_2.runtime_api_traits(),
+        RuntimeApiMetadata::hash,
+        RuntimeApiMetadata::hash,
+        RuntimeApiMetadata::name,
+    )
+}
+
+fn pallet_differences<'a>(
+    metadata_1: &'a Metadata,
+    metadata_2: &'a Metadata,
+) -> Vec<Diff<PalletMetadata<'a>>> {
+    diff(
+        metadata_1.pallets(),
+        metadata_2.pallets(),
+        PalletMetadata::hash,
+        PalletMetadata::hash,
+        PalletMetadata::name,
+    )
+}
+
+#[derive(Debug, Clone)]
+enum Diff<T> {
+    Added(T),
+    Changed { from: T, to: T },
+    Removed(T),
+}
+
+fn diff<T, C: PartialEq, I: Hash + PartialEq + Eq>(
```

Comment:
Awesome! Much prefer this :)

Reply:
(and with this logic now separated, you could also have a quick unit test or two that diffing things using this function works as expected :))

Reply:
I added a unit test. This also made me realize that the output order was random because of the random nature of hash maps. So I added alphabetical ordering as well.